### PR TITLE
Added allies-06a to the Red Alert mod

### PR DESCRIPTION
**Pull request by abc013**

### CODE REVIEW

**Comment by obrakmann** [comment]
Please add the .lua files to the `OpenRA.sln` file, for completeness' sake.

**Comment by abc013** [comment]
Updated.

**Comment by Mailaender** [comment]
You mix indentions with tabs and spaces in the .lua file, which makes it hard to read.

**Comment by abc013** [comment]
Crap, updated it.

File: mods/ra/maps/allies-06a/map.yaml [hunk 1]
```diff
@@ -0,0 +1,650 @@
+MapFormat: 11
+
+RequiresMod: ra
+
+Title: 06a: Iron Curtain research center
+# Infiltrate and destroy Iron Curtain research center
```
Comment (abcdefg30) [inline mods/ra/maps/allies-06a/map.yaml]: This needs to be resolved (the original title, commented out here, is too long for the UI).

File: mods/ra/maps/allies-06a/rules.yaml [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+MSUB:
+	Buildable:
+		Prerequisites: ~disabled
+
+SS:
+	AutoTarget:
+		InitialStanceAI: AttackAnything
+
+
```
Comment (abcdefg30) [inline mods/ra/maps/allies-06a/rules.yaml]: Duplicated new line.

File: mods/ra/maps/allies-06a/allies06a.lua [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+	end)
+end
+
+InfiltrateTechCenter = function()
+	Utils.Do(SovietTechLabs, function(a)
+		Trigger.OnInfiltrated(a, function()
+			DestroySovietsObj = player.AddPrimaryObjective("Destroy all Soviet buildings and units in the area.")
+			player.MarkCompletedObjective(InfiltrateTechCenterObj)
+
+			Proxy = Actor.Create( "powerproxy.paratroopers", false, { Owner = ussr, Location = CPos.New(0,0) })
```
Comment (abcdefg30) [inline mods/ra/maps/allies-06a/allies06a.lua]: This can be `local proxy`.

Comment (abcdefg30) [inline mods/ra/maps/allies-06a/allies06a.lua]: This can be `local proxy`. You also don't need to give it a `Location` init, afaik.

File: mods/ra/maps/allies-06a/allies06a.lua [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+	Patrol2Sub.Patrol({ SubPatrol2_1.Location, SubPatrol2_2.Location })
+	Patrol3Sub1.Patrol(SovietSubPath)
+	Patrol3Sub2.Patrol(SovietSubPath)
+end
+
+InitialAlliedReinforcements = function()
+	local camera = Actor.Create("Camera", true, { Owner = player, Location = DefaultCameraPosition.Location })
+	Trigger.AfterDelay(DateTime.Seconds(30), camera.Destroy)
+
+	Reinforcements.Reinforce(player, AlliedReinforcementsA, { AlliedEntry1.Location, UnitBStopLocation.Location}, 2)
```
Comment (abcdefg30) [inline mods/ra/maps/allies-06a/allies06a.lua]: Missing space before the `}`.

File: mods/ra/maps/allies-06a/allies06a.lua [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+	Trigger.AfterDelay(DateTime.Seconds(2), function()
+		Reinforcements.Reinforce(player, AlliedReinforcementsB, { AlliedEntry2.Location, UnitAStopLocation.Location}, 2)
```
Comment (abcdefg30) [inline mods/ra/maps/allies-06a/allies06a.lua]: Same here.

File: mods/ra/maps/allies-06a/allies06a.lua [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+	end)
+	Trigger.AfterDelay(DateTime.Seconds(5), function()
+		Reinforcements.Reinforce(player, { "mcv" }, { AlliedEntry3.Location, MCVStopLocation.Location})
```
Comment (abcdefg30) [inline mods/ra/maps/allies-06a/allies06a.lua]: And here.

**Comment by abc013** [comment]
Updated.

**Comment by abc013** [comment]
Updated, sry,  didn't notice that error 😕

File: mods/ra/maps/allies-06a/map.yaml (hunk 1 above)
Comment (GraionDilach) [inline mods/ra/maps/allies-06a/map.yaml]: "Sabotage the Iron Curtain research"?

Comment (GraionDilach) [inline mods/ra/maps/allies-06a/map.yaml]: "Cripple the Iron Curtain research"?

Comment (abc013) [inline mods/ra/maps/allies-06a/map.yaml]: Still too long, but only two letters less and it fits in.

Comment (GraionDilach) [inline mods/ra/maps/allies-06a/map.yaml]: Then just drop the "the" from it into "Cripple Iron Curtain research". :smiley:

Comment (pchote) [inline mods/ra/maps/allies-06a/map.yaml]: Why don't you just make the sidebar wider? RA's is ~40px narrower than TD.

Comment (abc013) [inline mods/ra/maps/allies-06a/map.yaml]: @pchote should i make an own PR for it?

Comment (pchote) [inline mods/ra/maps/allies-06a/map.yaml]: How many other long mission titles does RA have?  If this is a rare case then shortening the names is sufficient IMO.

Comment (abc013) [inline mods/ra/maps/allies-06a/map.yaml]: I think these titles are too long:
### Allied:
◾Infiltrate and destroy Iron Curtain research Center

◾Protect the Chronosphere research center

◾Secure the river for the navy

### Soviet:
◾Capture the control centers

**Comment by abc013** [comment]
updated.

File: mods/ra/maps/allies-06a/allies06a-AI.lua [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+		end
+	end
+
+	if target and yak.AmmoCount() > 0 then
+		yak.Attack(target)
+	else
+		yak.ReturnToBase()
+	end
+
+	yak.CallFunc(function()
```
Comment (GraionDilach) [inline mods/ra/maps/allies-06a/allies06a-AI.lua]: Is this intended to continuously loop recursively?

Comment (abcdefg30) [inline mods/ra/maps/allies-06a/allies06a-AI.lua]: Yes. And it kinda has a delay. The yak executes the `CallFunc` activity once the other activities are finished.

File: mods/ra/maps/allies-06a/allies06a.lua [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+
+	Trigger.OnCapture(Radar, function()
+		player.MarkCompletedObjective(CaptureRadarDomeObj)
+	end)
+end
+
+InfiltrateTechCenter = function()
+	Utils.Do(SovietTechLabs, function(a)
+		Trigger.OnInfiltrated(a, function()
+			DestroySovietsObj = player.AddPrimaryObjective("Destroy all Soviet buildings and units in the area.")
```
Comment (obrakmann) [inline mods/ra/maps/allies-06a/allies06a.lua]: I think there should be a check here whether another tech lab has already been infiltrated, otherwise you'd get duplicate objectives, and you'd be overwriting the reference to the earlier one(s) so you cannot mark them complete anymore.

**Review by obrakmann — CHANGES_REQUESTED** [review]
Looks overall fine, just a couple of minor things to safe-guard against players doing unexpected things.

File: mods/ra/maps/allies-06a/allies06a-AI.lua (hunk 7 above)
Comment (GraionDilach) [inline mods/ra/maps/allies-06a/allies06a-AI.lua]: Is that working properly even? From my tests, CallFunc breaks the activity queue and the sole reason this is unnoticable in say, ProductionAirdrop is because the actors affected there also has FlyAwayOnIdle.

**Comment by abc013** [comment]
Updated.

**Review by obrakmann — APPROVED** [review]

**Comment by reaperrr** [comment]
I got a crash pretty early:
´´´Fatal Lua Error: [string "BindingSupport.lua"]:30: Uncaught CLR exception at Lua->CLR boundary: System.Reflection.TargetInvocationException: Ein Aufrufziel hat einen Ausnahmefehler verursacht. ---> System.Reflection.TargetInvocationException: Ein Aufrufziel hat einen Ausnahmefehler verursacht. ---> Eluant.LuaException: Unknown actor type '1'
   bei OpenRA.Mods.Common.Scripting.ReinforcementsGlobal.CreateActor(Player owner, String actorType, Boolean addToWorld, Nullable`1 entryLocation, Nullable`1 nextLocation)
   bei OpenRA.Mods.Common.Scripting.ReinforcementsGlobal.ReinforceWithTransport(Player owner, String actorType, String[] cargoTypes, CPos[] entryPath, CPos[] exitPath, LuaFunction actionFunc, LuaFunction exitFunc)´´´

File: mods/ra/maps/allies-06a/allies06a-AI.lua [hunk 9]
```diff
@@ -0,0 +1,226 @@
+
+WTransWays =
+{
+	{ WaterUnloadEntry1.Location, WaterUnload1.Location },
+	{ WaterUnloadEntry2.Location, WaterUnload2.Location }
+}
+WTransUnits =
+{
+	hard = { { "3tnk", "3tnk", "3tnk", "v2rl", "v2rl" }, { "v2rl", "v2rl", "e4", "e4", "3tnk" } },
+	normal = { { "e1", "1", "3tnk", "3tnk", "v2rl" }, { "e4", "e4", "e4", "e4", "v2rl" } },
```
Comment (GraionDilach) [inline mods/ra/maps/allies-06a/allies06a-AI.lua]: Caused by this typo here.

**Comment by reaperrr** [comment]
In addition, I think the allied color could be bit more blueish and less white, the current color is a bit too close to the snow terrain, in my opinion.

**Comment by pchote** [comment]
>In addition, I think the allied color could be bit more blueish and less white, the current color is a bit too close to the snow terrain, in my opinion.

There's no reason why RA couldn't use the original indexed player colours, like we do in TD.

**Comment by reaperrr** [comment]
> There's no reason why RA couldn't use the original indexed player colours, like we do in TD.

Sure, but the original color scheme was somewhere between this mission and the shellmap, so it could use some tweaking either way.

**Comment by GraionDilach** [comment]
I'd say that the indexed player colors in the campaign should be set up in a separate PR though.

**Comment by Micr0Bit** [comment]
finished it on easy without any problems 

:+1:

**Comment by reaperrr** [comment]
> I'd say that the indexed player colors in the campaign should be set up in a separate PR though.

Yeah, I'm fine with merging this once the typo bug is fixed.

**Comment by abc013** [comment]
Updated it.

File: mods/ra/maps/allies-06a/allies06a-AI.lua (hunk 9 above)
Comment (abc013) [inline mods/ra/maps/allies-06a/allies06a-AI.lua]: Oops 😆

**Comment by reaperrr** [comment]
👍

**Comment by reaperrr** [comment]
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/c081343b5f1deccfd3f73bdb904046669257e962...d4c273e3f2cf1eeabba071fa063219c9871630ca)

**Comment by Biofreak1987** [comment]
I have no chance of leaving the coast. The enemy fills the sea with submarines and destroys all of my naval yards almost instantly. 

Is it supposed to be like this? What am I doing wrong?

File: mods/ra/maps/allies-06a/allies06a-AI.lua (hunk 7 above)
Comment (abcdefg30) [inline mods/ra/maps/allies-06a/allies06a-AI.lua]: It works. At least the last time I tried it (which is like a month ago).

**Comment by abcdefg30** [comment]
Sounds like you need to be faster in getting a shipyard and some boats. IIRC the AI will build waves of three submarines from time to time (they just shouldn't get the time to mass up).

**Comment by reaperrr** [comment]
I was too slow myself, but I 'solved' that by placing lots of e3 and arty at the beach to kill a bunch of them each time I placed a ship yard until they were all destroyed.

**Comment by pchote** [comment]
>   I 'solved' that by placing lots of e3 at the beach to kill a bunch of them each time I placed a ship yard until they were all destroyed.

This sounds horrible for (I assume) the default difficulty level.

**Comment by obrakmann** [comment]
The `ProduceNaval` function should probably only run after the player built a shipyard.

**Comment by Micr0Bit** [comment]
once you managed to build the shipyard in the little bay and a couple gunboats to deal with the first wave of subs all upcoming subs "stuck" anyway on the bay's corner ... and the torpedos only hit the corner instead of the shipyard ... gunboats and transporters then get ignored

i then did the same as reaperrr did , couple of e3 deal with every new sub near the coast

**Comment by Echelon9** [comment]
For what it's worth, I also initially got stuck getting waved by the subs in the gulf. Probably would have given up the mission, until realised just turtling on and expending a few sub yards is the way to go here...
